website_sale_product_contract_gift/models/sale_order.py

```diff
@@ -0,0 +1,181 @@
+# SPDX-FileCopyrightText: 2024 Coop IT Easy SC
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+
+import datetime
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    is_gift = fields.Boolean(compute="_compute_is_gift")
+    gift_date = fields.Date(compute="_compute_gift_date")
+
+    @api.depends("order_line")
+    def _compute_gift_date(self):
+        """Get gift_date from date_start on order_line"""
+        for order in self:
+            order.gift_date = order.order_line.filtered(
+                lambda r: r.product_id.is_gift
+            ).mapped("date_start")[0]
+
+    @api.depends("order_line")
+    def _compute_is_gift(self):
+        """Tell if an order is a gift or not"""
+        for order in self:
+            order.is_gift = any(order.order_line.mapped("product_id").mapped("is_gift"))
+
+    @api.constrains("order_line")
+    def _check_gift_alone(self):
+        """Ensure gift are not mixed in a sale order."""
+        for order in self:
+            if self.order_line:
+                at_least_one_gift = any(
+                    order.order_line.mapped("product_id").mapped("is_gift")
+                )
+                is_all_gift = all(
+                    order.order_line.mapped("product_id").mapped("is_gift")
+                )
+                if at_least_one_gift and not is_all_gift:
+                    raise ValidationError(
+                        _(
+                            "Cannot add product gift in an order that "
+                            "contains other product that are not gifts."
+                        )
+                    )
```

Comment:
avoid iterating twice over the list.
```suggestion
            if len(set(order.order_line.product_id.mapped("is_gift"))) > 1:
                raise ValidationError(
                    _(
                        "Cannot mix gift and non-gift products in the same order."
                    )
                )
```

Reply:
Nice one.

Reply:
not done yet.

Reply:
(solution stolen from [here](https://stackoverflow.com/a/3787948))